File: apps/mobile/test/widgets/friend_card_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/core/theme.dart';
import 'package:mobile/models/friendship.dart';
import 'package:mobile/widgets/friend_card.dart';

void main() {
  final friendship = FriendshipSummary(
    id: 'f1',
    status: FriendshipStatus.accepted,
    createdAt: DateTime(2026, 1, 1),
    user: const FriendUserSummary(pseudo: 'alice', avatarUrl: null),
  );

  testWidgets('kind friend affiche "Retirer" et déclenche onRemove', (
    tester,
  ) async {
    var removed = false;

    await tester.pumpWidget(
      MaterialApp(
        theme: buildReverbLightTheme(),
        home: Scaffold(
          body: FriendCard(
            friendship: friendship,
            kind: FriendCardKind.friend,
            onRemove: () async => removed = true,
          ),
        ),
      ),
    );

    expect(find.text('alice'), findsOneWidget);
    await tester.tap(find.text('Retirer'));
    await tester.pumpAndSettle();

    expect(removed, isTrue);
  });

  testWidgets('kind received affiche Accepter et Refuser', (tester) async {
    var accepted = false;

    await tester.pumpWidget(
      MaterialApp(
        theme: buildReverbLightTheme(),
        home: Scaffold(
          body: FriendCard(
            friendship: friendship,
            kind: FriendCardKind.received,
            onAccept: () async => accepted = true,
            onRemove: () async {},
          ),
        ),
      ),
    );

    expect(find.text('Accepter'), findsOneWidget);
    expect(find.text('Refuser'), findsOneWidget);

    await tester.tap(find.text('Accepter'));
    await tester.pumpAndSettle();

    expect(accepted, isTrue);
  });

  testWidgets('kind sent affiche "Demande envoyée" et un bouton Annuler', (
    tester,
  ) async {
    var cancelled = false;

    await tester.pumpWidget(
      MaterialApp(
        theme: buildReverbLightTheme(),
        home: Scaffold(
          body: FriendCard(
            friendship: friendship,
            kind: FriendCardKind.sent,
            onRemove: () async => cancelled = true,
          ),
        ),
      ),
    );

    expect(find.text('Demande envoyée'), findsOneWidget);
    await tester.tap(find.text('Annuler'));
    await tester.pumpAndSettle();

    expect(cancelled, isTrue);
  });
}

File: apps/web/src/lib/components/friend/FriendCard.test.ts
import { describe, expect, it, vi } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/svelte';
import type { FriendshipSummary } from '@reverb/shared';
import FriendCard from './FriendCard.svelte';

const friendship: FriendshipSummary = {
	id: 'f1',
	status: 'ACCEPTED',
	createdAt: '2026-01-01T10:00:00.000Z',
	user: { pseudo: 'bob', avatarUrl: null }
};

describe('FriendCard', () => {
	it('kind=friend : "Retirer" appelle onRemove', async () => {
		const onRemove = vi.fn().mockResolvedValue(undefined);
		render(FriendCard, { friendship, kind: 'friend', onRemove });

		await fireEvent.click(screen.getByRole('button', { name: 'Retirer' }));

		await waitFor(() => expect(onRemove).toHaveBeenCalled());
	});

	it('kind=received : "Accepter" appelle onAccept', async () => {
		const onAccept = vi.fn().mockResolvedValue(undefined);
		render(FriendCard, { friendship, kind: 'received', onAccept, onRemove: vi.fn() });

		await fireEvent.click(screen.getByRole('button', { name: 'Accepter' }));

		await waitFor(() => expect(onAccept).toHaveBeenCalled());
	});

	it('kind=sent : affiche "Demande envoyée" et "Annuler" appelle onRemove', async () => {
		const onRemove = vi.fn().mockResolvedValue(undefined);
		render(FriendCard, { friendship, kind: 'sent', onRemove });

		expect(screen.getByText('Demande envoyée')).toBeInTheDocument();

		await fireEvent.click(screen.getByRole('button', { name: 'Annuler' }));

		await waitFor(() => expect(onRemove).toHaveBeenCalled());
	});
});
